### PR TITLE
Significantly restructure torch/aten import design.

### PR DESCRIPTION
**Pull request by silvasean**
This is a really major and invasive restructuring of the way we get
torch operators (`torch::jit::Operator` / `c10::OperatorHandle`) into
MLIR. Please forgive the challenging review, but due to the sheer
invasiveness, it wasn't really practical do do it in sane smaller
pieces.

This fully replaces everything that was already working on the
TorchScript path (actually, more -- we added tanh support to
TorchToLinalg in order to delete the older code paths). Additionally,
I've kept the lights on for the acap path too, including what little e2e
stuff was working before (for expediency I made a few tiny compromises
along the way that will be easy to undo when we give that path proper
attention).

Overview of the new design:
- The torch operator `somens::someunqualname.someoverloadname` is
  imported as `torch.somens.someunqualname.someoverloadname` (skip the
  last dotted part if the overload name is empty), OR, if we don't have
  such an op registered, it is imported as
  `torch.operator "somens.someunqualname.someoverloadname" (...) : ...`.
  - The addition of the "overload name" is a critical element here, as
    the `(ns,unqual,overload)` triple is unique, which solves a lot of
    problems we were having.
  - This involves having separate MLIR ops for the `trailing_` and
    `.out` variants and all the different overloads. This seemed
    necessary, because the set of overloads is so wild and varied and
    unstructured. The previous design was leaning into some underlying
    structure that just isn't there -- the default situation is
    the "random overload that we want to manage on the MLIR side",
    rather than that being an exception. E.g.  `aten::ne` (not-equal)
    has 21 overloads, only 4 of which are c10 dispatcher ops see
    [gist](https://gist.github.com/silvasean/190ba918c550c956260e21254e1b8aa1),
    and the "out" variant is really called `.Tensor_out` instead of
    `.out` as it frequently is for other ops.
  - Rationale for all being in `torch` namespace: the set of operators
    are so varied and unstructured that "dialect per namespace"
    doesn't result in anything resembling the typical MLIR dialect
    boundary expectations. We could maybe draw the boundary at
    dispatcher ops vs non-dispatcher ops, but that doesn't seem to
    really result in very much useful structure at this point in time.
  - Note: within the torch operator registry, we effectively have a
    mini-basicpy subdialect (already type-resolved), which is reasonably
    structured.
  - The existing Torch op interfaces are also removed -- now that we
    track the overload name, we can losslessly find the original
    operator.
- Instead of `ATenRecognizeKernelsPass`, we now have a
  `ReduceOpVariantsPass` that keys off certain traits (and perhaps
  eventually interfaces) to reduce variants of ops to a smaller set,
  ideally operating on immutable tensors and using surrounding ops to
  model the mutability/aliasing aspects.
  - Note: `torch.ns.unqual.overload` ops allow both immutable and
    mutable tensors (unlike the previous hard distinction in the common
    case). This is a premonition for a future change that will introduce a
    bona fide `!torch.tensor` type that will clean up a bunch of stuff.
- `TorchToLinalg` / `TorchToStd` supercede the existing
  "ATen->TCF->TCP->Linalg" path.
- The new `torch_ods_gen.py` supercedes `torch_signature_ods_gen.py`.
  It should look somewhat familiar, but the benefit of hindsight has
  allowed a lot of simplifications.

The overall trend seems to be to make the `torch` dialect a nice layer
independent of anything else. It feels like as a natural result of
various future changes we will be removing the reliance on basicpy+numpy
dialects and have a nice self-contained type system too that properly
models the TorchScript type system (including proper subtyping,
mutable/immutable tensors, optional dtype, etc.).

Recommended review order:
- Start at some of the new import IR, e.g. in
  `frontends/pytorch/test/node_import/prim.py`,
  `frontends/pytorch/test/acap_export/test_export_add3.py`, and other
  tests.
- `frontends/pytorch/python/torch_mlir_utils/codegen/torch_ods_gen.py`
  and associated generated files:
  - `include/npcomp/Dialect/Torch/IR/GeneratedAtenOps.td`
  - `include/npcomp/Dialect/Torch/IR/GeneratedPrimOps.td`
- Inspect `ReduceOpVariants.cpp` / `reduce-op-variants.mlir` and the new
  traits in `include/npcomp/Dialect/Torch/IR/TorchTraits.h`
- Various code changes in the import path in
  `frontends/pytorch/csrc/builder`. Probably most interesting is the new
  code in `torch_to_mlir_utils.cpp` that has the logic to create the
  `torch.operator` ops or `torch.ns.unqual.overload` ops.

This is the [new ResNet IR](https://gist.github.com/silvasean/5407aafb710d07612b7b5b92eabecebe),
just to be able to look at a substantial sample of IR in the new style.

### CODE REVIEW

File: frontends/pytorch/csrc/builder/torch_to_mlir_utils.cpp [hunk 1]
```diff
@@ -339,3 +339,42 @@ torch_mlir::derefineValues(c10::ArrayRef<MlirValue> values,
...
+  auto separatorPosition = opNameSuffix.find_first_of("::");
+  assert(separatorPosition != std::string::npos);
+  opNameSuffix.replace(separatorPosition, 2, ".");
+  const std::string &overloadName = schema.overload_name();
+  if (!overloadName.empty()) {
+    opNameSuffix = opNameSuffix + "." + overloadName;
+  }
+  std::string opName = "torch." + opNameSuffix;
+  // If we have a registered op, use it!
+  if (mlirContextIsRegisteredOperation(context, toMlirStringRef(opName))) {
```
Comment (stellaraccident) [inline frontends/pytorch/csrc/builder/torch_to_mlir_utils.cpp:363]: Nice. I like this switch between opaque/not opaque.

File: lib/Dialect/Torch/IR/TorchOps.cpp [hunk 2]
```diff
@@ -204,6 +182,59 @@ void DerefineOp::getCanonicalizationPatterns(RewritePatternSet &patterns,
...
+  return nullptr;
+}
+
+//===----------------------------------------------------------------------===//
+// AtenLenTOp
+//===----------------------------------------------------------------------===//
+
+void AtenLenTOp::getCanonicalizationPatterns(RewritePatternSet &patterns,
+                                             MLIRContext *context) {
+  patterns.add(+[](AtenLenTOp op, PatternRewriter &rewriter) {
```
Comment (stellaraccident) [inline lib/Dialect/Torch/IR/TorchOps.cpp:208]: I don't believe the unary `+` is required for >C++11.

File: test/Conversion/TorchToLinalg/basic.mlir [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+func @torch.aten.mm$no_convert$missing_dtype(%arg0: tensor<*x!numpy.any_dtype>, %arg1: tensor<*x!numpy.any_dtype>) -> tensor<*x!numpy.any_dtype> {
+  // CHECK-NEXT: torch.aten.mm
+  %0 = torch.aten.mm %arg0, %arg1 : tensor<*x!numpy.any_dtype>, tensor<*x!numpy.any_dtype> -> tensor<*x!numpy.any_dtype>
+  return %0 : tensor<*x!numpy.any_dtype>
+}
+
+// CHECK-LABEL: func @torch.aten.mm$no_convert$wrong_rank
+func @torch.aten.mm$no_convert$wrong_rank(%arg0: tensor<?xf32>, %arg1: tensor<?xf32>) -> tensor<*x!numpy.any_dtype> {
+  // CHECK-NEXT: torch.aten.mm
+  %0 = torch.aten.mm %arg0, %arg1 : tensor<?xf32>, tensor<?xf32> -> tensor<*x!numpy.any_dtype>
```
Comment (stellaraccident) [inline test/Conversion/TorchToLinalg/basic.mlir:35]: Do you have this test case swapped with the description from the next one? (that one seems to be about rank, and this one seems to be a missing dtype variant, in fact)

File: frontends/pytorch/python/torch_mlir_utils/codegen/torch_ods_gen.py [hunk 4]
```diff
@@ -0,0 +1,481 @@
...
+        if has_canonicalizer:
+            p("let hasCanonicalizer = 1;")
+    p("}")
+    p("\n")
+
+
+def emit_op(operator: JitOperator,
+            f: TextIO,
+            *,
+            traits: List[str] = None,
```
Comment (stellaraccident) [inline frontends/pytorch/python/torch_mlir_utils/codegen/torch_ods_gen.py:336]: The type here needs to be `Optional[List[str]]`

File: frontends/pytorch/python/torch_mlir_utils/codegen/torch_ods_gen.py [hunk 5]
```diff
@@ -0,0 +1,481 @@
+#  Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
```
Comment (stellaraccident) [inline frontends/pytorch/python/torch_mlir_utils/codegen/torch_ods_gen.py:1]: Note to self: We should add a Github action which runs this at Torch head and verifies that no diffs were introduced.

**Review by stellaraccident — APPROVED** [review]
This is a *really* nice simplification/relayering. Very nice to see the representation collapsed into one well defined set of constructs in the torch dialect.

(and fwiw, I found it to be a comparatively simple and understandable diff, given the magnitude of the representation change)

File: lib/Dialect/Torch/IR/TorchOps.cpp (hunk 2 above)
Comment (silvasean) [reply lib/Dialect/Torch/IR/TorchOps.cpp:208]: Unfortunately it still seems to be needed on clang 9 on my machine. I think it has something to do with overload resolution / template matching and not the decay itself.

Comment (silvasean) [reply lib/Dialect/Torch/IR/TorchOps.cpp:208]: clang 11 (latest on my machine) also has the same issue btw :(

File: test/Conversion/TorchToLinalg/basic.mlir (hunk 3 above)
Comment (silvasean) [reply test/Conversion/TorchToLinalg/basic.mlir:35]: Good catch. Fixed and added explanatory comments. For this one the key thing is that the operands are rank 1 when the mm requires static rank 2.

File: frontends/pytorch/python/torch_mlir_utils/codegen/torch_ods_gen.py (hunk 5 above)
Comment (silvasean) [reply frontends/pytorch/python/torch_mlir_utils/codegen/torch_ods_gen.py:1]: Good idea. I can take a stab too to get my feet wet with CI stuff too.

**Pull request merged**